File: linked_list/reverse_nodes/25_reverse_nodes_in_k_group.cc
#include <tuple>
#include <utility>

struct ListNode {
  int val;
  ListNode* next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode* next) : val(x), next(next) {}
};

// 2024/01/14: 直接抄答案
// 翻转一个子链表，并且返回新的头与尾
std::pair<ListNode*, ListNode*> MyReverse(ListNode* head, ListNode* tail) {
  ListNode* prev = tail->next;
  ListNode* curr = head;
  while (prev != tail) {
    ListNode* next = curr->next;
    curr->next = prev;
    prev = curr;
    curr = next;
  }
  return {tail, head};
}

ListNode* ReverseKGroup(ListNode* head, int k) {
  ListNode* dummy_node = new ListNode(0);
  dummy_node->next = head;
  ListNode* prev = dummy_node;

  while (head) {
    ListNode* tail = prev;
    // 查看剩余部分长度是否大于等于 k
    for (int i = 0; i < k; ++i) {
      tail = tail->next;
      if (!tail) {
        return dummy_node->next;
      }
    }
    ListNode* next = tail->next;
    // 这里是 C++17 的写法，也可以写成
    // std::pair<ListNode*, ListNode*> result = MyReverse(head, tail);
    // head = result.first;
    // tail = result.second;
    std::tie(head, tail) = MyReverse(head, tail);
    // 把子链表重新接回原链表
    prev->next = head;
    tail->next = next;
    prev = tail;
    head = tail->next;
  }

  return dummy_node->next;
}

File: linked_list/reverse_nodes/25_reverse_nodes_in_k_group.py
from linked_list.list_node import ListNode, create_list_node, print_list_node
from typing import Optional


def revers_list(head: Optional[ListNode]) -> Optional[ListNode]:
    prev = None
    curr = head
    while curr:
        next = curr.next
        curr.next = prev
        prev = curr
        curr = next
    return prev


# 2026/05/14: 自己实现, 采用双重递归, 下次可以尝试 reverse_k_group 使用循环实现.
def reverse_k_group(head: Optional[ListNode], k: int) -> Optional[ListNode]:
    if head is None:
        return None

    curr = head

    for i in range(k - 1):
        curr = curr.next
        if curr is None:
            return head
    next_group = curr.next
    curr.next = None
    reversed_list = revers_list(head)
    head.next = reverse_k_group(next_group, k)
    return reversed_list


head = [1, 2, 3, 4, 5]
k = 2
head = create_list_node(head)
print_list_node(reverse_k_group(head, k))
